src/v/cloud_storage/remote_segment.cc

```diff
@@ -177,6 +182,56 @@ remote_segment::data_stream(size_t pos, ss::io_priority_class io_priority) {
     co_return data_stream;
 }
 
+ss::future<> remote_segment::do_hydrate() {
+    auto callback = [this](
+                      uint64_t size_bytes,
+                      ss::input_stream<char> s) -> ss::future<uint64_t> {
+        offset_index tmpidx(get_base_rp_offset(), get_base_kafka_offset(), 0);
+        auto [sparse, sput] = input_stream_fanout<2>(std::move(s), 1);
+        auto parser = make_remote_segment_index_builder(
+          std::move(sparse), tmpidx, _base_offset_delta, 1); // TODO: use 64KB
+        auto fparse = parser->consume().finally(
+          [parser] { return parser->close(); });
+        auto fput = _cache.put(_path, sput).finally([sref = std::ref(sput)] {
+            return sref.get().close();
+        });
+        auto [rparse, rput] = co_await ss::when_all(
```

Comment:
I noticed recently that there is a `ss::coroutine::all` helper and I'm not sure if it offers something unique over co_awaiting on ss::when_all.